### PR TITLE
mlpack: new package

### PR DESCRIPTION
**Pull request by wdconinc**
mlpack is an intuitive, fast, and flexible header-only C++ machine learning library with bindings to other languages. It is meant to be a machine learning analog to LAPACK, and aims to implement a wide array of machine learning methods and functions as a "swiss army knife" for machine learning researchers.

### CODE REVIEW

**Comment by alecbcs** [comment]
@wdconinc thanks for the PR! This looks great! Just checking, any interest in maintaining this package?

**Comment by wdconinc** [comment]
Sure, I can add myself as maintainer. We're evaluating this for inclusion in our stack, so I'm hardly an expert user (and not at all of the Go and Julia layers). It's just often easier these days to write a spack package...

File: var/spack/repos/builtin/packages/mlpack/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/mlpack/package.py:18]:
```suggestion

maintainers("wdconinc")

```

File: var/spack/repos/builtin/packages/mlpack/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/mlpack/package.py]:
```suggestion
    maintainers("wdconinc")

```

File: var/spack/repos/builtin/packages/mlpack/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/mlpack/package.py:22]: Stop eating trailing blank lines...
```suggestion

    version("4.1.0", sha256="e0c760baf15fd0af5601010b7cbc536e469115e9dd45f96712caa3b651b1852a")
```

File: var/spack/repos/builtin/packages/mlpack/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/mlpack/package.py:22]:
```suggestion
    version("4.2.0", sha256="f780df984a71029e62eeecdd145fb95deb71b133cefc7840de0ec706d116dd60")
    version("4.1.0", sha256="e0c760baf15fd0af5601010b7cbc536e469115e9dd45f96712caa3b651b1852a")
```
Changelog: https://github.com/mlpack/mlpack/compare/4.1.0...4.2.0

No other changes needed for 4.2.0.

**Comment by wdconinc** [comment]
Status update on this PR:
- Go bindings: commented out due to my inability to get a `gonum` package working...
- Julia bindings: untested due to my inability to get an llvm compiled with the julia patches...
- Python bindings: fail to install due to `pip install` step that expects `pip` in the Python dist-packages... (issue filed upstream)
- R bindings: this actually works :-)

File: var/spack/repos/builtin/packages/mlpack/package.py
Comment (wdconinc) [inline var/spack/repos/builtin/packages/mlpack/package.py:52]:
```suggestion
        depends_on("py-pip")
        depends_on("py-wheel")
```

**Comment by wdconinc** [comment]
This now builds both the R and Python bindings correctly (and the julia bindings probably too, but I need a working julia spack instance). I think it is ready for merging since this now allows for C++, R, and Python interfaces, which meets the necessary threshold in my opinion.

@alecbcs Can you give it another review?

**Comment by wdconinc** [comment]
Note to self: don't rename branch before it has been merged...

**Review by becker33 — APPROVED** [review]

**Pull request merged**
